src/Client/HedgedConnections.cpp

```diff
@@ -0,0 +1,416 @@
+            default:
+                /// If we receive an exception or an unknown packet, we save it.
+                res = std::move(packet);
+                break;
+        }
+    }
+
+    return res;
+}
+
+Packet HedgedConnections::receivePacket()
+{
+    std::lock_guard lock(cancel_mutex);
+    return receivePacketUnlocked({});
+}
+
+Packet HedgedConnections::receivePacketUnlocked(AsyncCallback async_callback)
+{
+    if (!sent_query)
+        throw Exception("Cannot receive packets: no query sent.", ErrorCodes::LOGICAL_ERROR);
+    if (!hasActiveConnections())
+        throw Exception("No more packets are available.", ErrorCodes::LOGICAL_ERROR);
+
+    if (epoll.empty())
+        throw Exception("No pending events in epoll.", ErrorCodes::LOGICAL_ERROR);
+
+    return receivePacketImpl(std::move(async_callback));
+}
+
+Packet HedgedConnections::receivePacketImpl(AsyncCallback async_callback)
+{
+    int event_fd;
+    ReplicaStatePtr replica = nullptr;
+    Packet packet;
+    bool finish = false;
+    while (!finish)
+    {
+        event_fd = getReadyFileDescriptor(async_callback);
+
+        if (fd_to_replica.find(event_fd) != fd_to_replica.end())
+        {
+            replica = fd_to_replica[event_fd];
+            packet = receivePacketFromReplica(replica, async_callback);
+            finish = true;
+        }
+        else if (timeout_fd_to_replica.find(event_fd) != timeout_fd_to_replica.end())
+        {
+            replica = timeout_fd_to_replica[event_fd];
+            processTimeoutEvent(replica, replica->active_timeouts[event_fd]);
+        }
+        else if (event_fd == get_hedged_connections.getFileDescriptor())
+            tryGetNewReplica();
+        else
+            throw Exception("Unknown event from epoll", ErrorCodes::LOGICAL_ERROR);
+    }
+
+    return packet;
+};
+
+int HedgedConnections::getReadyFileDescriptor(AsyncCallback async_callback)
+{
+    for (auto & [fd, replica] : fd_to_replica)
+        if (replica->connection->hasReadPendingData())
+            return replica->fd;
+
+    return epoll.getReady(std::move(async_callback)).data.fd;
+}
+
+Packet HedgedConnections::receivePacketFromReplica(ReplicaStatePtr & replica, AsyncCallback async_callback)
+{
+    Packet packet = replica->connection->receivePacket(std::move(async_callback));
+    switch (packet.type)
+    {
+        case Protocol::Server::Data:
+            removeTimeoutsFromReplica(replica, epoll, timeout_fd_to_replica);
+            processReceiveData(replica);
+            addTimeoutToReplica(TimerTypes::RECEIVE_TIMEOUT, replica, epoll, timeout_fd_to_replica, get_hedged_connections.getConnectionTimeouts());
+            break;
+        case Protocol::Server::Progress:
+        case Protocol::Server::ProfileInfo:
+        case Protocol::Server::Totals:
+        case Protocol::Server::Extremes:
+        case Protocol::Server::Log:
+            removeTimeoutFromReplica(TimerTypes::RECEIVE_TIMEOUT, replica, epoll, timeout_fd_to_replica);
```

Comment:
I think it is not quite correct to remove timeout only after we got packet.
Cause if we are reading some packet very slow, it may cause timer alert (even if data is still sending).